src/utils/alias_method.cpp

```diff
@@ -0,0 +1,87 @@
+//========================================================================================
+// (C) (or copyright) 2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+//! \file alias_method.cpp
+//  \brief Construct tables used to sample from a discrete probability distribution using
+//  the alias method
+
+#include "alias_method.hpp"
+
+#include <queue>
+
+namespace parthenon {
+namespace AliasMethod {
+
+AliasMethod::AliasMethod(const std::vector<Real> &probabilities)
+    : prob_table("prob_table", probabilities.size()),
+      alias_table("alias_table", probabilities.size()) {
+
+  // make probability table and alias table, this implementation is based on
+  // https://gist.github.com/Liam0205/0b5786e9bfc73e75eb8180b5400cd1f8
+  auto host_prob_table = Kokkos::create_mirror_view(prob_table);
+  auto host_alias_table = Kokkos::create_mirror_view(alias_table);
+
+  Real prob_sum = 0.0;
+  for (size_t i = 0; i < probabilities.size(); ++i) {
+    // printf("p[%lu] = %.8e\n", i, probabilities[i]);
+    prob_sum += probabilities[i];
+  }
+
+  std::queue<int> under_full, over_full;
+  for (size_t i = 0; i < probabilities.size(); ++i) {
+    host_alias_table(i) = -1;
+
+    host_prob_table(i) = Real(probabilities.size()) * probabilities[i] / prob_sum;
+    if (host_prob_table(i) < 1.0)
+      under_full.push(i);
+    else
+      over_full.push(i);
```

Comment:
This is more stylistic but I prefer putting opening and closing braces to prevent accidentally shifting lines and changing logic.

Reply:
Even better:
```cxx
(host_prob_table(i) < 1.0 ? under_full : over_full).push(i);
```

Reply:
under_full is missing a push but ya that is clear.